Fix BIG iso data path

Thalley opened the pull request:
The iso data path can only be set for one direction when dealing with BISs, so the CIS way of setting the data path can not be fully reused. 

Also adds a small update to the BIGinfo reports in the shell.

Thalley committed:
Bluetooth: ISO: Fix issue with setting iso data path for BISs

When setting the iso data path for a BIS, it shall only be set for
one direction (depending on whether we are a broadcaster or
a receiver).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: shell: Mark BIGinfo reports more clearly in the shell

The BIGinfo reports were almost identical to the
periodic advertising reports when printet. Updated to
mark the start of the line with BIGinfo.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Vudentz commented:
Does the spec says that BT_ISO_DATA_PATH_DISABLED is not required for BIS though? We always do this for CIS since it is not clear what the controller would assume to be the default path if nothing is programed, but if we can assume the default is disabled then we wouldn't need to send it for CIS as well.

Thalley commented:
> Does the spec says that BT_ISO_DATA_PATH_DISABLED is not required for BIS though? We always do this for CIS since it is not clear what the controller would assume to be the default path if nothing is programed, but if we can assume the default is disabled then we wouldn't need to send it for CIS as well.

@Vudentz 

From the core spec: 
> If the Host attempts to set an output data path using a connection handle that is for an Isochronous Broadcaster, for an input data path on a Synchronized Host Controller Interface Functional Specification Receiver, or for a data path for the direction on a unidirectional CIS where BN is set to 0, the Controller shall return the error code Command Disallowed (0x0C).

So it's not allowed to set the data path for
1) Output (controller to host) for a broadcaster
2) Input (host to controller) for a receiver
3) For a unidirectional CIS where BN is set to 0 (I guess we are still missing a check for this).

Thalley commented:
Found missing handling for removing the ISO path. Will update ASAP.

Thalley committed:
Bluetooth: ISO: Only remove a single data path for broadcasters

Since we are only setting up a single data path for
broadcasters (either receive or transmit), we should
only remove that same path.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -761,6 +777,28 @@ static void bt_iso_remove_data_path(struct bt_conn *conn)
+
+		hci_le_remove_iso_data_path(conn, dir);
+		return;
+	} /* else CIS: remove both directions */
```
Maybe just have an else? Seems simpler to me.

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -705,29 +705,45 @@ static int bt_iso_setup_data_path(struct bt_conn *conn)
+		}
+
+	} else {
+		err = hci_le_setup_iso_data_path(conn, &in_path);
```
Add same comment as below for removing.
` /* CIS: add both directions */`

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -705,29 +705,45 @@ static int bt_iso_setup_data_path(struct bt_conn *conn)
+	rx_qos = chan->qos->rx;
 
-	if (!chan->qos->tx) {
+	in_path.path = tx_qos && tx_qos->path ? tx_qos->path : &path;
```
Can the path variable be renaming to something that is easier to understand?
disabled_path perhaps?

Looks like initializing it to BT_ISO_DATA_PATH_DISABLED could make some of the code below simpler as well.

Thalley replied:
If I understand it correctly, then the `path` variable isn't a diabled path, but rather a default (HCI) path. 

There are some parts of this that I'm not 100% sure about, so we'd have to get @Vudentz input, but will at least rename `path` to `default_path`.

joerchan replied:
But if tx_qos is NULL then it is set to PATH_DISABLED below, I'm not sure what it is set to if rx_qos->path is NULL though.

Thalley replied:
It's confusing to me as well as both `struct bt_iso_data_path` and `struct bt_iso_chan_path` has a path ID (pid). @Vudentz Do you have any input here?

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -705,29 +705,45 @@ static int bt_iso_setup_data_path(struct bt_conn *conn)
-		return err;
-	}
+	if (conn->iso.is_bis) {
+		if (tx_qos) {
```
For the unitiated the direction of the path and it's relation to in/out path looks strange, is this behavior properly documented in the API?

Thalley replied:
That's a very good question. I think it's fairly well documented that if `tx` is NULL then you cannot transmit, and if `rx` is NULL you cannot receive, but if you think there is something missing there, please let us know.

joerchan replied:
I don't understand the behavior when the application does not initialize the path in QoS.

```
	/** Channel data path reference.
	 *  Setting to NULL default to HCI data path.
	 */
	struct bt_iso_chan_path		*path;
```

What does "default to HCI data path" mean?

The other thing is the how the NULL of tx_qos/rx_qos behaves for BIS.
The code sets up only one direction based off TX_QOS is NULL or not.
So if both tx_qos and rx_qos is set only in_path is set.
Is there something preventing this from being the case?
Should that be clarified in the documentation?

Thalley replied:
> What does "default to HCI data path" mean?
Regarding the data path and HCI: The data path is either HCI, where data is then sent to the host via HCI, or a proprietary data path where the data isn't handled over HCI. The application can either provide a HCI path, a proprietary path or NULL to default to the HCI path. 

> The code sets up only one direction based off TX_QOS is NULL or not.
> So if both tx_qos and rx_qos is set only in_path is set.
The API only allows either `rx` or `tx` to be set for a broadcast ISO, so it shouldn't be confusing for the application. 

> Is there something preventing this from being the case?
The spec also only allows setting a single data path (for either IN or OUT depending on the BIS), so that's why we only set one. 

> Should that be clarified in the documentation?
No, I don't think that was very clear. Added more text for the `rx` and `tx` fields.

Thalley committed:
Bluetooth: ISO: Improved text for setting ISO data paths

Improved the documentation for the ISO data paths, as well
as making the bt_iso_setup_data_path function a bit more
clear.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
@joerchan I've addressed your comments as best I can. I agree that there are still some things that definitely needs improvements, but I'd much rather deal with those in a separate PR as this PR fixes some critical issues with creating a BIG, and I think the remaining issues are outside the scope of this PR.

Vudentz commented on subsys/bluetooth/host/iso.c:
```diff
@@ -777,9 +779,33 @@ static void bt_iso_remove_data_path(struct bt_conn *conn)
+		struct bt_iso_chan_io_qos *tx_qos;
+		uint8_t dir;
+
+		chan = SYS_SLIST_PEEK_HEAD_CONTAINER(&conn->channels, chan, node);
```
Usually the remove_data_path is only called when all channels have been removed, or this is not the case for BIS? Anyway it might be better that we store the direction(s) within conn->iso then to avoid such lookups so both BIS and CIS should become pretty similar if we can tell what data path have been programmed.

Thalley replied:
For BIS they act more as a group: All BIS are either created or removed at the same time (i.e. you create or terminate a BIG (or a BIG sync), rather than individual BISes).

Thalley replied:
@Vudentz Just did a test for 2 BIS in a BIG and this works as intended :)

joerchan commented on include/bluetooth/iso.h:
```diff
@@ -33,6 +33,8 @@ extern "C" {
 				  BT_HCI_ISO_HDR_SIZE + \
 				  BT_HCI_ISO_DATA_HDR_SIZE)
 
+#define BT_ISO_DATA_PATH_HCI     0x00
```
Add a brief?

joerchan commented on subsys/bluetooth/host/iso.c:
```diff
@@ -700,34 +700,52 @@ static int bt_iso_setup_data_path(struct bt_conn *conn)
+			return err;
+		}
+
+		return hci_le_setup_iso_data_path(conn, &out_path);
```
Is it a problem if setting up the out-path fails and the in-path succeeded, would it then be needed to remove the in-path again?

Thalley replied:
If this function fails, we disconnect the ISO connection. We automatically remove all data paths in the disconnect callback. 

The only thing that maybe need to be improved, is that we don't check if a data path has been setup before attempting to remove it (e.g. in case this fails), so if this fails, then remove data path will also fail (we ignore errors on that at the moment though). 

I'd rather leave a comment to get that fixed in another PR, so we can fix the issues as hand first.

Thalley commented:
@Vudentz Any additional input? 
Otherwise, can we get this merged @carlescufi ?
